### DISCUSSION TITLE
One process, many stores?

### DISCUSSION THREAD

**Discussion by jeffparsons** [bytecodealliance/wrpc#122] [General]
Hello! 👋 

Do you imagine wRPC as (eventually) being the best way to make calls between component instances from different Wasmtime stores _within the same process_? Or would you imagine something built deeper into Wasmtime could be meaningfully more efficient? 

For my own uses I currently have a generic thunk thingy that maps values between the types in different stores and dynamically defines functions on linkers to facilitate cross-store calls (it's trivial — doesn't handle resources at all because I don't need that), but I imagine there's a lot of information that Wasmtime has internally that it could use to skip a bunch of that work, e.g. prove type equality once, and just copy bytes forever after.

Asking here because it looks like you've recently been deep within the bowels of Wasmtime!

Not asking for concrete answers here — just seeking general thoughts on the idea of overheads of different approaches and where wRPC sits in the solution space. 🙂

**Comment by rvolosatovs** [comment]
In-process "transport" is certainly a possibility. In terms of overhead it's hard to say without looking at the implementation - in theory the runtime could optimize calls between components and pass raw canonical ABI value bytes around directly, which would eliminate the need for any encoding/decoding.
In practice, however, I am not aware of a better way to dynamically satisfy imports other than using `Linker::func_new` APIs - since that all is based on the inefficient `Val` wrapper, I would guess that codec overhead would be miniscule in comparison and so I think it's unlikely that there'd be any performance penalty using wRPC.
On a related note, there's hoping that we'd be able to get efficient value wrappers from the runtime, which would provide an optimized way of getting a hold of their "value definition"-encoded views, a few of these conversions between the canon ABI and value definition encoding are actually just an identity function (bools, results and options are encoded the same, for example)

If we really wanted *the best* way to call components in-process, we'd probably want to pass "pointers" around directly instead of dealing with bytes, that would have to be built in Wasmtime itself and is probably out of scope for this project.
